st2actions/st2actions/runners/ssh/parallel_ssh.py

```diff
@@ -0,0 +1,317 @@
+        """
+
+        options = {
+            'path': path,
+            'force': force
+        }
+        return self._execute_in_pool(self._delete_dir, **options)
+
+    def close(self):
+        """
+        Close all open SSH connections to hosts.
+        """
+
+        for host in self._hosts_client.keys():
+            try:
+                self._hosts_client[host].close()
+            except:
+                LOG.exception('Failed shutting down SSH connection to host: %s', host)
+
+    def _execute_in_pool(self, execute_method, **kwargs):
+        results = {}
+
+        for host in self._bad_hosts.keys():
+            results[host] = self._bad_hosts[host]
+
+        for host in self._hosts_client.keys():
+            while not self._pool.free():
+                eventlet.sleep(self._scan_interval)
+            self._pool.spawn(execute_method, host=host, results=results, **kwargs)
+
+        self._pool.waitall()
+        return results
+
+    def _connect(self, host, results, raise_on_error=False):
+        (hostname, port) = self._get_host_port_info(host)
+
+        extra = {'_host': host, '_port': port, '_user': self._ssh_user}
+        if not self._ssh_password:
+            extra['_key'] = self._ssh_key
+        else:
+            extra['_password'] = '<redacted>'
+        LOG.debug('Connecting to host.', extra=extra)
+
+        client = ParamikoSSHClient(hostname, username=self._ssh_user,
+                                   password=self._ssh_password,
+                                   key=self._ssh_key, port=port)
+        try:
+            client.connect()
+        except:
+            error = 'Failed connecting to host %s.' % hostname
+            LOG.exception(error)
+            _, ex, tb = sys.exc_info()
+            if raise_on_error:
+                raise
+            error = ' '.join([self.CONNECT_ERROR, str(ex)])
+            error_dict = self._generate_error_result(error, tb)
+            self._bad_hosts[hostname] = error_dict
+            results[hostname] = error_dict
+        else:
+            self._hosts_client[hostname] = client
+            results[hostname] = {'message': 'Connected to host.'}
+
+    def _run_command(self, host, cmd, results, timeout=None):
+        try:
+            LOG.debug('Running command: %s on host: %s.', cmd, host)
+            client = self._hosts_client[host]
+            (stdout, stderr, exit_code) = client.run(cmd, timeout=timeout)
+            is_succeeded = (exit_code == 0)
+            results[host] = {'stdout': stdout, 'stderr': stderr, 'return_code': exit_code,
+                             'succeeded': is_succeeded, 'failed': not is_succeeded}
+        except:
+            error = 'Failed executing command %s on host %s', cmd, host
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _put_files(self, local_path, remote_path, host, results, mode=None,
```

Comment:
Just to confirm it - `remote_path` is always a randomly generated directory in `/tmp`, right?

Because if we use the same path across multiple invocations there are potential races and other issues.